subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectConfigurationReportingGradle.kt

```diff
@@ -0,0 +1,346 @@
+    }
+
+    override fun afterProject(action: Action<in Project>) {
+        delegate.afterProject(action.withCrossProjectModelAccessCheck())
+    }
+
+    override fun getDefaultProject(): ProjectInternal = crossProjectModelAccess.access(referrerProject, delegate.defaultProject)
+
+    override fun getGradle(): Gradle = this
+
+    override fun addListener(listener: Any) {
+        delegate.addListener(maybeWrapListener(listener))
+    }
+
+    override fun removeListener(listener: Any) {
+        delegate.removeListener(maybeWrapListener(listener))
+    }
+
+    override fun equals(other: Any?): Boolean =
+        javaClass == (other as? CrossProjectConfigurationReportingGradle)?.javaClass &&
+            other.delegate == delegate &&
+            other.referrerProject == referrerProject
+
+    override fun hashCode(): Int = Objects.hash(delegate, referrerProject)
+
+    override fun toString(): String = "CrossProjectConfigurationReportingGradle($delegate)"
+
+    internal
+    companion object {
+        fun from(gradle: GradleInternal, referrerProject: ProjectInternal): CrossProjectConfigurationReportingGradle {
+            val parentCrossProjectModelAccess = gradle.services.get(CrossProjectModelAccess::class.java)
+            val parentCrossProjectConfigurator = gradle.services.get(CrossProjectConfigurator::class.java)
+            return CrossProjectConfigurationReportingGradle(gradle, referrerProject, parentCrossProjectModelAccess, parentCrossProjectConfigurator)
+        }
+    }
+
+    private
+    fun maybeWrapListener(listener: Any): Any = when (listener) {
+        is ProjectEvaluationListener -> CrossProjectModelAccessProjectEvaluationListener(listener, referrerProject, crossProjectModelAccess)
+        // all the supported listener types other than ProjectEvaluationListener are already reported as configuration cache problems in non-buildSrc builds
+        else -> listener
+    }
+
+    private
+    fun <T> Closure<T>.withCrossProjectModelAccessChecks(): Closure<T> =
+        CrossProjectModelAccessTrackingClosure(this, referrerProject, crossProjectModelAccess)
+
+    private
+    fun Action<in Project>.withCrossProjectModelAccessCheck(): Action<Project> {
+        val originalAction = this@withCrossProjectModelAccessCheck
+        return Action<Project> {
+            val originalProject = this@Action
+            check(originalProject is ProjectInternal) { "Expected the projects in the model to be ProjectInternal" }
+            originalAction.execute(crossProjectModelAccess.access(referrerProject, originalProject))
+        }
+    }
+
+    private
+    fun Action<in Gradle>.withCrossProjectModelGradleAccessCheck(): Action<Gradle> {
+        val originalAction = this@withCrossProjectModelGradleAccessCheck
+        return Action<Gradle> {
+            val originalGradle = this@Action
+            check(originalGradle is GradleInternal) { "Expected the projects in the model to be ProjectInternal" }
```

Comment:
🟡 Something like _Expected Gradle instance to be GradleInternal_?

Reply:
Fixed it, thank you!

Reply:
Fixed in a fixup commit, thanks!